### PR TITLE
rshiny

### PR DESCRIPTION
**Pull request by ms4952**
Created a Dashboard in R using Shiny package.

### CODE REVIEW

File: examples/r-shiny/app.R [hunk 1]
```diff
@@ -0,0 +1,31 @@
+library(ggplot2)
```
Comment (jnolis) [inline examples/r-shiny/app.R:1]: I rewrote this for a couple of reasons (we can talk about that separately):

```r
library(ggplot2)
library(shiny)

data <- read.csv("https://saturn-public-data.s3.us-east-2.amazonaws.com/examples/dashboard/housePriceData.csv")
model <- lm(SalePrice~BedroomAbvGr+YearBuilt, data)

ui <- fluidPage(
    titlePanel("Housing Data"),
    sidebarLayout(
        sidebarPanel(
          sliderInput("bedrooms", "Number of Bedrooms", min=0,max=8, value=3,step=1)
        ),
        mainPanel(
           plotOutput("price_by_year")
        )
    )
)
    
server <- function(input, output) {
    
    output$price_by_year <- renderPlot({
      data <- data.frame(YearBuilt = 1970:2015)
      data$BedroomAbvGr <- input$bedrooms
      data$prediction <- predict(model, newdata = data)
      
      ggplot(data, aes(x = YearBuilt, y = prediction))+
        geom_col(fill = "#FF6721") +
                   theme_minimal() +
        scale_y_continuous(labels = scales::dollar) +
        labs(x = "Year home built", y = "Predicted value of house")
    })
}
shinyApp(ui = ui, server = server)
```

File: examples/r-job/.saturn/saturn.json [hunk 2]
```diff
@@ -2,7 +2,7 @@
   "name": "example-r-job",
   "image_uri": "saturncloud/saturn-rstudio:2021.11.10",
   "description": "Run a job on Saturn Cloud using R .",
-  "working_directory": "/home/jovyan/git-repos/examples/examples/r-job",
+  "working_directory": "/home/jovyan/examples/examples/r-job",
```
Comment (jnolis) [inline examples/r-job/.saturn/saturn.json:5]: Can you remove this and make it a separate PR?

File: examples/r-shiny/.saturn/saturn.json [hunk 3]
```diff
@@ -0,0 +1,16 @@
+{
+  "name": "example-r-shiny",
+  "image_uri": "saturncloud/saturn-rstudio:2021.11.10",
+  "description": "Deploy a Dashboard to Saturn Cloud using R Shiny .",
```
Comment (jnolis) [inline examples/r-shiny/.saturn/saturn.json]:
```suggestion
  "description": "Deploy a dashboard to Saturn Cloud using R Shiny",
```

File: examples/r-shiny/README.md [hunk 4]
```diff
@@ -0,0 +1,125 @@
+---
+title: "Create a Dashboard with Shiny"
+linkTitle: "Create a Dashboard with Shiny"
+weight: 3
+description: Use Shiny to create simple Dashboard and deploy it in Saturn Cloud
+---
+## What is a Dashboard?
+Dashboards is a tool where you can view you raw data in form of meaningful visualisations. It transforms data to charts, graphs and tables that help in analysis of information, defining KPIs  and eventually taking data centric decisions. 
+## What is Shiny?
```
Comment (jnolis) [inline examples/r-shiny/README.md]: Unlike an API, I think we can assume people know what a dashboard is
```suggestion
```

File: examples/r-shiny/README.md [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+<a href="https://shiny.rstudio.com/" target='_blank' rel='noopener'>Shiny</a> is a widely used package for building interactive web applications in R.  Shiny is easy to use and flexible. It integrates nicely with CSS themes, htmlwidgets and JavaScript actions.
+You can create dashboards or run machine learning models through web apps etc without any prior knowledge of web development tools, using Shiny package.
```
Comment (jnolis) [inline examples/r-shiny/README.md]:
```suggestion
You can create dashboards or run machine learning models through interactive applications without any prior knowledge of web development tools, using the R Shiny package. 
```

File: examples/r-shiny/README.md [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+---
+## What is a Dashboard?
+Dashboards is a tool where you can view you raw data in form of meaningful visualisations. It transforms data to charts, graphs and tables that help in analysis of information, defining KPIs  and eventually taking data centric decisions. 
+## What is Shiny?
+<a href="https://shiny.rstudio.com/" target='_blank' rel='noopener'>Shiny</a> is a widely used package for building interactive web applications in R.  Shiny is easy to use and flexible. It integrates nicely with CSS themes, htmlwidgets and JavaScript actions.
+You can create dashboards or run machine learning models through web apps etc without any prior knowledge of web development tools, using Shiny package. 
+Shiny app structure constitutes : 
+
+**User Interface**: Responsible for creating the layout of your web app, like panels, textbox, dropdown etc. This is basically what users sees at their end.\
+**Server**: Constitutes logic which will display content in the webapp as per the inputs given by user from UI.  \
```
Comment (jnolis) [inline examples/r-shiny/README.md]:
```suggestion
**User Interface**: Responsible for creating the layout of your web app, like panels, textbox, dropdown etc. This is basically what users sees at their end.
**Server**: Constitutes logic which will display content in the webapp as per the inputs given by user from UI. 
```

File: examples/r-shiny/README.md [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+**ShinyApp**: This function launches the shiny app as per UI/server pair.
+
+## Objective
+In the example below we are building a dashboard as R script (`.R` file) using Shiny. For this example we are creating a histogram for sale prices as per selected number of bins. The data for same is taken from [Kaggle](https://www.kaggle.com/c/house-prices-advanced-regression-techniques/data).
+
+## Building a Dashboard
+First we will create an R script which installs shiny and ggplot2 packages from CRAN. Let's name this file as `setup.R`.
```
Comment (jnolis) [inline examples/r-shiny/README.md]: ggplot2 is included in the `saturncloud-rstudio` image, so I would mention that and remove the explicit installation of ggplot2

File: examples/r-shiny/README.md [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+User will select a number as single input from slider This number represents value of bins for histogram plot  In the following code in first line we have loaded the ggplot2 and shiny packages. 
+Now we will read house price data in variable data.
+
+In UI function we will create appearance of our dashboard. Title of our dashboard "Housing Data" is added to title panel.
+We then give structure to our app using sidebarLayout function. sidebarLayout contains sidebarPanel and mainPanel. In slider input we are setting number of bins ranging from 1 to 30.
+Since we want our output to show histogram plot, we will use function plotOutput inside mainPanel. 
+plotOutput displays outputs returned from render functions (renderPlot in our case). 
+
+In server function we have logic to create histogram as per selected number of bins from slider. Function renderPlot will return output to function plotOutput. 
+In our case renderPlot fetches qplot object and stores it in variable `distPlot`. From housing dataset we first extract sale prices of houses. Then generate bins as per the user input and then draw histogram based on bin number using qplot function.
```
Comment (jnolis) [inline examples/r-shiny/README.md]: This should be rewritten for the updated shiny app

File: examples/r-shiny/README.md [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+We then give structure to our app using sidebarLayout function. sidebarLayout contains sidebarPanel and mainPanel. In slider input we are setting number of bins ranging from 1 to 30.
+Since we want our output to show histogram plot, we will use function plotOutput inside mainPanel. 
+plotOutput displays outputs returned from render functions (renderPlot in our case). 
+
+In server function we have logic to create histogram as per selected number of bins from slider. Function renderPlot will return output to function plotOutput. 
+In our case renderPlot fetches qplot object and stores it in variable `distPlot`. From housing dataset we first extract sale prices of houses. Then generate bins as per the user input and then draw histogram based on bin number using qplot function. 
+
+In the last line we are call shinyApp and pass in UI and server variables.
+
+```R
```
Comment (jnolis) [inline examples/r-shiny/README.md:36]: Switch to the new code I included

File: examples/r-shiny/README.md [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+```
+
+## Deployment
+
+Save all three files (`setup.R`, `app.R`, `run-app.R`) in a directory inside GitHub repo. Now click **New Deployment**. It can be found on top right side of the resource page. 
+
+![deploy](https://saturn-public-assets.s3.us-east-2.amazonaws.com/example-resources/plumber_deployment.png "doc-image")
+
+
+To run, add following to command field.
```
Comment (jnolis) [inline examples/r-shiny/README.md]:
```suggestion
To run, add following to the **command** field of the deployment.
```

File: examples/r-shiny/README.md [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+```bash
+Rscript run-app.R
+```
+Add path to file in working directory field as shown below. In the screenshot below, my file `run-app.R` is inside repository Dashboard, directory shiny. If you aren't familiar with how to set up SSH credentials and add git repositories to Saturn Cloud check [here](https://saturncloud.io/docs/using-saturn-cloud/gitrepo/).
+
+![deploy](https://saturn-public-assets.s3.us-east-2.amazonaws.com/example-resources/Shiny-deploy.png "doc-image")
```
Comment (jnolis) [inline examples/r-shiny/README.md]: This needs to be updated to be the correct path for the example.

File: examples/r-shiny/README.md [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+Save all three files (`setup.R`, `app.R`, `run-app.R`) in a directory inside GitHub repo. Now click **New Deployment**. It can be found on top right side of the resource page. 
+
+![deploy](https://saturn-public-assets.s3.us-east-2.amazonaws.com/example-resources/plumber_deployment.png "doc-image")
+
+
+To run, add following to command field.
+```bash
+Rscript run-app.R
+```
+Add path to file in working directory field as shown below. In the screenshot below, my file `run-app.R` is inside repository Dashboard, directory shiny. If you aren't familiar with how to set up SSH credentials and add git repositories to Saturn Cloud check [here](https://saturncloud.io/docs/using-saturn-cloud/gitrepo/).
```
Comment (jnolis) [inline examples/r-shiny/README.md]: Moved the SSH stuff to where we mention git

```suggestion
Add path to file in working directory field as shown below--this should be the location off the `app.R` file.
```

File: examples/r-shiny/README.md [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+
+```
+Now create another R script which will run above piece of code. Let's name this as `run-app.R`. Start the server using shiny object. Since the host listens on 0.0.0.0, it will be reachable on an appropriate interface address to the connection. The API must use port 8000 for Saturn Cloud to be able to direct traffic to it.
+```R
+shiny::runApp(host="0.0.0.0", port=8000, launch.browser=FALSE)
+```
+
+## Deployment
+
+Save all three files (`setup.R`, `app.R`, `run-app.R`) in a directory inside GitHub repo. Now click **New Deployment**. It can be found on top right side of the resource page.
```
Comment (jnolis) [inline examples/r-shiny/README.md]: Moved the setting up git stuff from below

```suggestion
Save all three files (`setup.R`, `app.R`, `run-app.R`) in a directory inside GitHub repo. You'll need to connect [git on Saturn Cloud](https://saturncloud.io/docs/using-saturn-cloud/gitrepo/) to your git repo host. Now click **New Deployment**. It can be found on top right side of the resource page. 
```

File: examples/r-shiny/README.md [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+
+![script](https://saturn-public-assets.s3.us-east-2.amazonaws.com/example-resources/shiny-sh.png "doc-image")
+
+These are the rest of the fields you'll need to fill out
+
+
+**Name** - the name of the resource, which also is used as the URL for the Dashboard\
+**Instance Count** - How many concurrent machines to host the API on. Set to default as 1.\
+**Hardware and Size** - This example is small hence I have selected smallest available CPU resource.\
+**Image** - set to `saturn-rstudio` image, which is preconfigured to have R and some common R libraries.\
```
Comment (jnolis) [inline examples/r-shiny/README.md]: I don't think you need the backslashes on lists
```suggestion
**Instance Count** - How many concurrent machines to host the API on. Set to default as 1.
**Hardware and Size** - This example is small hence I have selected smallest available CPU resource.
**Image** - set to `saturn-rstudio` image, which is preconfigured to have R and some common R libraries.
```

File: examples/r-shiny/README.md [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+
+With this you are ready to deploy your dashboard by pressing the green button on the resource page of the deployment. 
+
+## Access Deployed Dashboard
+
+Click the URL given in deployment detail page . You will be redirected to Dashboard page. 
+
+![doc plumber](https://saturn-public-assets.s3.us-east-2.amazonaws.com/example-resources/shiny-dashboard.png "doc-image")
+
+Set the slider to the value you want for number of bins for histogram. By default number of bins in this example is set to 15. Histogram for housing sale prices is generated as per the selected value.
```
Comment (jnolis) [inline examples/r-shiny/README.md]: I don't think you need to explain how to use the app

```suggestion
```

File: examples/r-shiny/README.md [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+
+## Access Deployed Dashboard
+
+Click the URL given in deployment detail page . You will be redirected to Dashboard page. 
+
+![doc plumber](https://saturn-public-assets.s3.us-east-2.amazonaws.com/example-resources/shiny-dashboard.png "doc-image")
+
+Set the slider to the value you want for number of bins for histogram. By default number of bins in this example is set to 15. Histogram for housing sale prices is generated as per the selected value. 
+
+
```
Comment (jnolis) [inline examples/r-shiny/README.md]: Can you add something about publishing the app to everyone on the internet vs the default of just people logged in?

**Review by jnolis — CHANGES_REQUESTED** [review]
I have some writing changes and I adjusted the core Shiny app!

File: examples/r-job/.saturn/saturn.json (hunk 2 above)
Comment (ms4952) [reply examples/r-job/.saturn/saturn.json:5]: I haven't removed this since otherwise it will throw validation error.

**Review by jnolis — APPROVED** [review]
